sdk/spring/spring-cloud-azure-stream-binder-servicebus/src/main/java/com/azure/spring/cloud/stream/binder/servicebus/ServiceBusMessageChannelBinder.java

```diff
@@ -278,6 +278,9 @@ private ServiceBusTemplate getServiceBusTemplate() {
                 instrumentationManager.addHealthInstrumentation(instrumentation);
             });
             this.serviceBusTemplate = new ServiceBusTemplate(factory);
+            if (producerProperties.getExtension() != null) {
```

Comment:
I am wondering why this entity type is missing from the producer properties retrieved from line 271, the getProducerPropertiesSupplier?

Reply:
It's because the key to return the extended properties is going to the destination name. When it uses the Dynamic Destination name to pull the properties to create the producer it's not going to find it. Right here:
[Permalink](https://github.com/Azure/azure-sdk-for-java/blob/3a3499bfa1f28f52e6dce7aaf000b40dbb047216/sdk/spring/spring-messaging-azure-servicebus/src/main/java/com/azure/spring/messaging/servicebus/core/DefaultServiceBusNamespaceProducerFactory.java#L77)

```java
@Override
    public ServiceBusSenderAsyncClient createProducer(String name, ServiceBusEntityType entityType) {
        ProducerProperties producerProperties = this.propertiesSupplier.getProperties(name) != null
            ? this.propertiesSupplier.getProperties(name) : new ProducerProperties();
        if (entityType != null) {
            producerProperties.setEntityType(entityType);
        }
        return doCreateProducer(name, producerProperties);
    }
```

So it'll create a default `ProducerProperty` and if the `defaultEntityType` is set, it'll set the `entityType`. As the `defaultEntityType` is not set, then it'll fail to publish to the dynamic destination. That's why in there I'm not setting the `entityType` and setting the `defaultEntityType` in case the destination is dynamic is there is no ProducerProperty register and its name.

I've tested in our internal connector and it works fine when `defaultEntityType` is set.

Reply:
I see, thanks. So a better way to set the type is probably to use another header.

Reply:
Could you help add some comments around this code? Thanks.

Reply:
@saragluna 
So you mean you would like to add a new producer property so the user can define in the YAML what would be the defaultEntityType?
About the code, do you mean adding some comments in the code or just in the PR so it makes easier to understand? If it's the comments in the code, wouldn't be better some Java Docs instead?

Reply:
> So you mean you would like to add a new producer property so the user can define in the YAML what would be the defaultEntityType?
No, my original thought isn't correct. 

But I double checked the code, from AbstractMessageChannelBinder:doBindProducer -> ProvisioningProvider:provisionProducerDestination -> AbstractMessageChannelBinder:createProducerMessageHandler, it creates a producer. But in the case of dynamic binding, it will follow this path to resolve the binding, StreamBridge:resolveDestination -> BindingService:getExtendedProducerProperties (it will bind to the default properties) -> NewDestinationBindingCallback:configure, so we can configure the entity type for dynamic destionation via registering NewDestinationBindingCallback beans.

> About the code, do you mean adding some comments in the code or just in the PR so it makes easier to understand? If it's the comments in the code, wouldn't be better some Java Docs instead?
Yes, comment or Java Docs are both fine.

Reply:
I'll add the JavaDocs.

Regarding the [Dynamic Binding](https://docs.spring.io/spring-cloud-stream/reference/spring-cloud-stream/event-routing.html#spring-cloud-stream-sendto-destination), I've tried this approach but didn't work.

What I'm proposing is more similar to the [Dynamic Producer Destination we have on Solace Binder](https://github.com/SolaceProducts/solace-spring-cloud/tree/master/solace-spring-cloud-starters/solace-spring-cloud-stream-starter#dynamic-producer-destinations) behavior which can be easily achieve just by setting a `defaultEntityType` .

Basically right now the process is:
On `ServiceBusMessageChannelBinder:createProducerMessageHandler` is stores the `ProducerProperties` on the `extendedProducerPropertiesMap` with the destination as key -> `DefaultMessageHandler:handleMessageInternal:line65` it checks if there's a value in the header `azure_name` if yes it uses it if not it uses the destination -> `ServiceBusTemplate:sendAsync` -> `DefaultServiceBusNamespaceProducerFactory:createProducer` it tries to get the `ProducerProperty` from `extendedProducerPropertiesMap` which it'll not get it if you set the `azure_name` because the key was the original destination and in this case it creates a default `ProducerDestination` and if the `defaultEntityType` is set it'll set the default `ProducerDestination` with it.
Right now because the `defaultEntityType` is not set it fails on `DefaultServiceBusNamespaceProducerFactory:doCreateProducer`.

`azure_name` already works as the Dynamic Producer Destination header just like the [Dynamic Producer Destination we have on Solace Binder](https://github.com/SolaceProducts/solace-spring-cloud/tree/master/solace-spring-cloud-starters/solace-spring-cloud-stream-starter#dynamic-producer-destinations), the only issue is the missing `defaultEntityType`.

Reply:
Thanks for the update, the other way to set the entity type via the header I was talking about is exactly like this `setHeader(SolaceBinderHeaders.TARGET_DESTINATION_TYPE, "topic")`.

Reply:
I'll try the NewDestinationBindingCallback to see what causes the issue.

Reply:
Hi @Aryelmr, I've created this project https://github.com/saragluna/demo-servicebus-binder-dynamic-destinations to test the dynamic destinations. In my test, there are two ways to configure the default entity type:
  - Via the property `spring.cloud.stream.servicebus.default.producer.entityType`, see https://github.com/saragluna/demo-servicebus-binder-dynamic-destinations/blob/main/src/main/resources/application.properties
  - Via the [NewDestinationBindingCallback](https://github.com/saragluna/demo-servicebus-binder-dynamic-destinations/blob/3b5c3767fc3f43a51fdea0540d64a92bbf058033/src/main/java/com/example/demo_servicebus_binder/FunctionConfiguration.java#L21-L27). 

Could you help look to see whether these two approaches work in your case? If so, I think we don't need to make changes in this PR now. Thanks in advance.

Reply:
@saragluna 
Thank you for coming up with those alternatives to see if we could get the same results with the code changes :)
Unfortunately, they didn't work. 
On `DefaultServiceBusNamespaceProducerFactory:createProducer` it just creates a plain ProducerProperties so it doesn't get the `ExtendendProducerProperties` to get the default values and the NewDestinationBindingCallback didn't work as well. 

For the `setHeader(SolaceBinderHeaders.TARGET_DESTINATION_TYPE, "topic")` Solace approach is nice.
In the same way as the destination we could check if the EntityType header is set and if so pass the value. E.g.:

```
    @Override
    public <U> Mono<Void> sendAsync(String destination, Message<U> message) {
        Assert.hasText(destination, "destination can't be null or empty");
        ServiceBusEntityType entityType = getEntityType(message);
        ServiceBusSenderAsyncClient senderAsyncClient =
                     this.producerFactory.createProducer(destination, entityType);
        ServiceBusMessage serviceBusMessage = messageConverter.fromMessage(message, ServiceBusMessage.class);
        return senderAsyncClient.sendMessage(serviceBusMessage);
    }
    
    private String getEntityType(Message<?> message) {
        if (message.getHeaders().containsKey(SolaceBinderHeaders.TARGET_DESTINATION_TYPE)) {
            return message.getHeaders().get(SolaceBinderHeaders.TARGET_DESTINATION_TYPE, String.class);
        }

        return this.defaultEntityType;
    }
```

What do you think?

Reply:
Thanks. I like the header approach, and it looks better. But I am still wondering what your dynamic destination code looks like, and why it goes to the `DefaultServiceBusNamespaceProducerFactory:createProducer`? Could you help provide some code, so I could take a look at it.

Reply:
@saragluna 
I didn't get it. The binder already uses a header to get a Dynamic Destination when it uses the method `toDestination`. It checks the header and if the header has a value it uses the value as the destination. The issue is, today if you use it fails because it doesn't have a defaultEntityType set so when creating the ProducerProperties it fails. This is the part of code that it checks the header for a Dynamic Destination:
```java
private String toDestination(Message<?> message) {
        if (message.getHeaders().containsKey(AzureHeaders.NAME)) {
            return message.getHeaders().get(AzureHeaders.NAME, String.class);
        }

        return this.destination;
    }
```
 My changes is to fix the EntityType when using the Dynamic Destination the binder has today.

Regarding the changes, do you want to go with the Solace Binder Dynamic Destination approach and have a dynamic header to set the EntityType as well? If so I can make the implementation so you can review.

Additionally, when I deliver the changes for the dynamic header to set the EntityType how long until it gets to binder release?

We're using the ServiceBus Binder for how products and depending in the estimated time to get the changes into the binder we'll fork it. If you can provide any estimated time it'd be great for us to evaluate the effort of forking it vs waiting in getting the change into the binder.

Reply:
Hi @Aryelmr, the reason I asked how your dynamic destination is used is that I want to figure out why it's calling `ServiceBusMessageChannelBinder:createProducerMessageHandler` -> `DefaultMessageHandler:handleMessageInternal` without an `ExtendedProducerProperties`, which in my test it will always generate an `ExtendedProducerProperties` at least with default value. 

For the release question, if any PR gets in, it will be out with our monthly release, which normally happens at the beginning of each month.

Reply:
@saragluna 
Can you detail how you're reproducing using the dynamic destination?
Because if I set the `azure_name` header with a value for a destination the calling stack that happens here is this one I left in the [comment](https://github.com/Azure/azure-sdk-for-java/pull/39839#discussion_r1612226912).

And the reason it doesn't create a `ExtendedProducerProperties` is that when it doesn't find the ProducerProperties attached to a previous register Destination Name in the it just creates a plain ProducerProperties. You can see it right [here](https://github.com/Azure/azure-sdk-for-java/blob/3a3499bfa1f28f52e6dce7aaf000b40dbb047216/sdk/spring/spring-messaging-azure-servicebus/src/main/java/com/azure/spring/messaging/servicebus/core/DefaultServiceBusNamespaceProducerFactory.java#L78).

Reply:
Hi @Aryelmr, this is what I tried with https://github.com/saragluna/demo-servicebus-binder-dynamic-destinations/blob/3b5c3767fc3f43a51fdea0540d64a92bbf058033/src/main/java/com/example/demo_servicebus_binder/FunctionConfiguration.java#L34.

Reply:
Unfortunately this approach you're suggesting doesn't work in my case. The approach you sent in the last message it's to use the [Dynamic Binding](https://docs.spring.io/spring-cloud-stream/reference/spring-cloud-stream/event-routing.html#spring-cloud-stream-sendto-destination), which doesn't work for me because I'm using this binder to build a connector using the [Solace Connector Framework](https://mvnrepository.com/artifact/com.solace.connector.core/pubsubplus-connector-framework) which is framework on top of Spring Cloud Stream.

The approach that would work is the Solace Dynamic Producer Destination approach I mentioned, where we use the Headers to dynamic set the Destination and/or Entity Type or to set a default entity type and use the current "Dynamic Destination header" Azure binder already has.